### PR TITLE
[Feature] Feature NPUProfilerHook

### PR DESCRIPTION
**Pull request by luomaoling**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Added NPUProfilerHook class.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
No.

## Use cases (Optional)

```python
cfg = Config.fromfile(args.config) 
profiler_config = [dict(type='NPUProfilerHook', result_path='./cann_profiling')]
cfg.merge_from_dict({'custom_hooks': profiler_config })
runner = Runner.from_cfg(cfg)
```

Export timeline data
python /usr/local/Ascend/ascend-toolkit/latest/toolkit/tools/profiler/profiler_tool/analysis/msprof/msprof.py export  timeline -dir ./cann_profiling

Export summary data
python /usr/local/Ascend/ascend-toolkit/latest/toolkit/tools/profiler/profiler_tool/analysis/msprof/msprof.py export  summary -dir ./cann_profiling

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmengine/hooks/profiler_hook.py [hunk 1]
```diff
@@ -230,3 +232,94 @@ def _export_chrome_trace(self, runner):
...
+        end_process(bool): Whether to end the process after profiling.
+        use_e2e_profiler(bool): Turn on E2E profiling.
+        ge_profiling_to_std_out(bool): Turn on GE profiling.
+    Example:
+        >>> cfg = ...
+        >>> profiler_config = dict(type='NPUProfilerHook', end_process=True)
+        >>> cfg.merge_from_dict({'custom_hooks': custom_hooks})
+        >>> runner = Runner.from_cfg(cfg)
+        >>> runner.train()
+    """
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:263]:
```suggestion
    """NPUProfiler to analyze performance during training.

    NPU Profiler is a tool to analyze key performance bottlenecks in each
    stage of engineering running on Ascend AI processors during the training
    phase and make recommendations for performance optimization to achieve
    the ultimate performance of the product. The torch_npu.npu.profile
    interface is used to complete the profiling data collection at each
    stage of the project, and the data is analyzed by the msprof tool and
    the data can be dumped to further manually analyze the key performance
    bottlenecks. For more details on the torch_npu.npu.profile interface,
    please visit
    https://gitee.com/ascend/pytorch/blob/master/torch_npu/npu/profiler.py#profile

    Args:
        profile_start_iters (int): Number of start iterations for profiling.
            Defaults to 0.
        profile_end_iters (int): Number of end iterations for profiling.
            Defaults to 0.
        profiler_result_path (str): The path to save the profiling results file.
            Defaults to 'cann_profiling'.
        end_process (bool): Whether to end the process after profiling.
            Defaults to True.
        use_e2e_profiler (bool): Turn on E2E profiling. Defaults to False.
        ge_profiling_to_std_out (bool): Turn on GE profiling. Defaults to False.

    Example:
        >>> cfg = ...
        >>> profiler_config = dict(type='NPUProfilerHook', end_process=True)
        >>> cfg.merge_from_dict({'custom_hooks': custom_hooks})
        >>> runner = Runner.from_cfg(cfg)
        >>> runner.train()
    """
```

File: mmengine/hooks/profiler_hook.py [hunk 2]
```diff
@@ -230,3 +232,94 @@ def _export_chrome_trace(self, runner):
...
+                 end_process: bool = True,
+                 use_e2e_profiler: bool = False,
+                 ge_profiling_to_std_out: bool = False):
+
+        try:
+            import torch_npu
+        except ImportError:
+            raise ImportError('Failed to import torch_npu module')
+
+        if profile_start_iters > profile_end_iters:
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:279]:
```suggestion
        if profile_start_iters >= profile_end_iters:
```

File: mmengine/hooks/profiler_hook.py [hunk 3]
```diff
@@ -230,3 +232,94 @@ def _export_chrome_trace(self, runner):
...
+    @master_only
+    def after_train_iter(self,
+                         runner,
+                         batch_idx,
+                         data_batch=None,
+                         outputs=None):
+
+        if runner.iter == self.profile_end_iters:
+            runner.logger.info('profiler may take a few minutes...')
+            profiler_exit_result = self.profiler.__exit__(None, None, None)
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:323]: Is `profiler_exit_result` a boolean?

File: mmengine/hooks/profiler_hook.py (hunk 1 above)
Comment (zhouzaida) [reply mmengine/hooks/profiler_hook.py:263]: What are the `e2e` and `ge`?

File: mmengine/hooks/profiler_hook.py (hunk 3 above)
Comment (luomaoling) [reply mmengine/hooks/profiler_hook.py:323]: Yes.

File: mmengine/hooks/profiler_hook.py (hunk 2 above)
Comment (luomaoling) [reply mmengine/hooks/profiler_hook.py:279]: If it is >=, only two or more step profiling data can be collected by default, and in many cases, we still want to be able to collect only one step profiling data.

File: mmengine/hooks/profiler_hook.py (hunk 1 above)
Comment (luomaoling) [reply mmengine/hooks/profiler_hook.py]: e2e and ge related descriptive information has been added, thank you!

File: mmengine/hooks/profiler_hook.py (hunk 3 above)
Comment (luomaoling) [reply mmengine/hooks/profiler_hook.py:332]: Yes

File: mmengine/hooks/profiler_hook.py [hunk 4]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+    interface, please visit
+    https://gitee.com/ascend/pytorch/blob/master/torch_npu/npu/profiler.py#profile
+
+    Args:
+        profile_start_iters (int): Number of start iterations for profiling.
+            Defaults to 0.
+        profile_end_iters (int): Number of end iterations for profiling.
+            Defaults to 0.
+        profiler_result_path (str): The path to save the profiling
+            results file.Defaults to 'cann_profiling'.
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:255]:
```suggestion
        begin (int): Step at which to start profiling. Defaults to 0.
        end (int): Step at which to stop profiling. Defaults to 1.
        result_path (str): The path to save the profiling results file.
            Defaults to 'cann_profiling'.
```

File: mmengine/hooks/profiler_hook.py [hunk 5]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+        end_process (bool): Whether to end the process after profiling.
+            Defaults to True.
+        use_e2e_profiler (bool): Turn on E2E profiling, E2E profiling combines
+            performance data at the Pytorch level and the NPU level to analyze
+            the bottlenecks of model performance end-to-end, and cannot show
+            detailed content, and only as an auxiliary analysis.
+            Defaults to False.
+        ge_profiling_to_std_out (bool): Turn on GE profiling, GE uses to
+            collect the profiling data of the host side scheduling of the
+            Assend device.Defaults to False.
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:265]:
```suggestion
            Assend device. Defaults to False.
```

File: mmengine/hooks/profiler_hook.py [hunk 6]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+    @master_only
+    def after_train_iter(self,
+                         runner,
+                         batch_idx,
+                         data_batch=None,
+                         outputs=None):
+
+        if runner.iter == self.profile_end_iters:
+            runner.logger.info('profiler may take a few minutes...')
+            profiler_exit_result = self.profiler.__exit__(None, None, None)
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:332]:
```suggestion
            exit_status = self.profiler.__exit__(None, None, None)
```

File: mmengine/hooks/profiler_hook.py [hunk 7]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+    https://gitee.com/ascend/pytorch/blob/master/torch_npu/npu/profiler.py#profile
+
+    Args:
+        profile_start_iters (int): Number of start iterations for profiling.
+            Defaults to 0.
+        profile_end_iters (int): Number of end iterations for profiling.
+            Defaults to 0.
+        profiler_result_path (str): The path to save the profiling
+            results file.Defaults to 'cann_profiling'.
+        end_process (bool): Whether to end the process after profiling.
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:256]:
```suggestion
        exit_after_profiling (bool): Whether to exit the program after profiling.
```

File: mmengine/hooks/profiler_hook.py [hunk 8]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+
+    @master_only
+    def after_train_iter(self,
+                         runner,
+                         batch_idx,
+                         data_batch=None,
+                         outputs=None):
+
+        if runner.iter == self.profile_end_iters:
+            runner.logger.info('profiler may take a few minutes...')
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:331]:
```suggestion
            runner.logger.info('profiler may take a few minutes to save the profiling result.')
```

File: mmengine/hooks/profiler_hook.py [hunk 9]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+        if not osp.exists(self.result_path):
+            os.makedirs(self.result_path, exist_ok=True)
+
+        self.profiler = torch_npu.npu.profile(
+            self.result_path, use_e2e_profiler=use_e2e_profiler)
+
+    @master_only
+    def before_run(self, runner):
+
+        if self.begin - self.end > runner.max_iters:
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:309]: Why do you subtract `self.end` from `self.begin`?

File: mmengine/hooks/profiler_hook.py [hunk 10]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+                         runner,
+                         batch_idx,
+                         data_batch=None,
+                         outputs=None):
+
+        if runner.iter == self.end - 1:
+            runner.logger.info('profiler may take a few minutes to'
+                               ' save the profiling result.')
+            exit_status = self.profiler.__exit__(None, None, None)
+            if not exit_status and self.exit_after_profiling:
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:331]: Should we use `or`?

File: mmengine/hooks/profiler_hook.py [hunk 11]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+
+        try:
+            import torch_npu
+        except ImportError:
+            raise ImportError('Failed to import torch_npu module')
+
+        if begin >= end:
+            raise ValueError(
+                'The profile end iters should not be greater than or '
+                'equal to profile start iters')
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:290]:
```suggestion
                'The iteration to start profiling should not be greater than or '
                'equal to the end')
```

File: mmengine/hooks/profiler_hook.py [hunk 12]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+
+        self.profiler = torch_npu.npu.profile(
+            self.result_path, use_e2e_profiler=use_e2e_profiler)
+
+    @master_only
+    def before_run(self, runner):
+
+        if self.end > runner.max_iters:
+            raise ValueError('The profiling end iter should not be greater'
+                             'than the max epoch')
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:311]:
```suggestion
            raise ValueError('The profiling end iteration should not be greater'
                             'than the max iteration')
```

File: mmengine/hooks/profiler_hook.py [hunk 13]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+
+    @master_only
+    def before_train_iter(self, runner, batch_idx, data_batch=None):
+
+        if runner.iter == self.begin:
+            self.profiler.__enter__()
+            runner.logger.info('profiler is profiling...')
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:318]:
```suggestion
            runner.logger.info('NPUProfiler starts profiling...')
```

File: mmengine/hooks/profiler_hook.py [hunk 14]
```diff
@@ -230,3 +232,101 @@ def _export_chrome_trace(self, runner):
...
+        use_e2e_profiler (bool): Turn on E2E profiling, E2E profiling combines
+            performance data at the Pytorch level and the NPU level to analyze
+            the bottlenecks of model performance end-to-end, and cannot show
+            detailed content, and only as an auxiliary analysis.
+            Defaults to False.
+        ge_profiling_to_std_out (bool): Turn on GE profiling, GE uses to
+            collect the profiling data of the host side scheduling of the
+            Assend device. Defaults to False.
+
+    Example:
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:265]:
```suggestion
    Examples:
```

File: mmengine/hooks/profiler_hook.py [hunk 15]
```diff
@@ -230,3 +232,103 @@ def _export_chrome_trace(self, runner):
...
+    The torch_npu.npu.profile interface is used to complete the profiling data
+    collection at each stage of the project, and the data is analyzed by the
+    msprof tool and the data can be dumped to further manually analyze the
+    key performance bottlenecks. For more details on the torch_npu.npu.profile
+    interface, please visit
+    https://gitee.com/ascend/pytorch/blob/master/torch_npu/npu/profiler.py#profile
+
+    Args:
+        begin (int): Number of start iterations for profiling. Defaults to 0.
+        end (int): Number of end iterations for profiling. Defaults to 0.
```
Comment (zhouzaida) [inline mmengine/hooks/profiler_hook.py:251]:
```suggestion
        end (int): Number of end iterations for profiling. Defaults to1.
```

File: tests/test_hooks/test_profiler_hook.py [hunk 16]
```diff
@@ -202,3 +204,74 @@ def test_with_runner(self):
...
+        runner.max_iters = 10000
+        runner.logger = MMLogger.get_instance('test_npu_profiler')
+
+        runner.iter = 0
+
+        hook = NPUProfilerHook(begin=0, end=10, result_path=result_path)
+        hook.before_run(runner)
+
+        hook.profiler = MagicMock()
+        hook.after_train_iter(runner, 1, 1, 1)
```
Comment (zhouzaida) [inline tests/test_hooks/test_profiler_hook.py:247]:
```suggestion
        hook.after_train_iter(runner, 1)
```

File: tests/test_hooks/test_profiler_hook.py [hunk 17]
```diff
@@ -202,3 +204,74 @@ def test_with_runner(self):
...
+        runner = self.build_runner(self.epoch_based_cfg)
+        runner.train()
+
+        self.assertTrue(
+            ops.exists(result_path),
+            'ERROR::profiler result path is not generated!')
+
+        self.assertTrue(
+            os.getenv('GE_PROFILING_TO_STD_OUT', '0') == '1',
+            'ERROR::GE PROFILING fail!')
```
Comment (zhouzaida) [inline tests/test_hooks/test_profiler_hook.py:277]:
```suggestion
            'ERROR::profiler result path is not generated!')

        self.assertTrue(
            os.getenv('GE_PROFILING_TO_STD_OUT', '0') == '1',
            'GE PROFILING fail!')
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**